src/storage/src/hummock/compactor/fast_compactor_runner.rs

```diff
@@ -687,31 +687,27 @@ impl<F: TableBuilderFactory> CompactTaskExecutor<F> {
             self.may_report_process_key(1);
 
             let mut drop = false;
-            let epoch = iter.key().epoch_with_gap.pure_epoch();
             let value = HummockValue::from_slice(iter.value()).unwrap();
-            if is_new_user_key || self.last_key.is_empty() {
+            let is_first_or_new_user_key = is_new_user_key || self.last_key.is_empty();
+            if is_first_or_new_user_key {
                 self.last_key.set(iter.key());
-                self.watermark_can_see_last_key = false;
                 self.last_key_is_delete = false;
             }
 
             // See note in `compactor_runner.rs`.
-            if epoch < self.task_config.watermark
+            if !self.task_config.retain_multiple_version
                 && self.task_config.gc_delete_keys
                 && value.is_delete()
             {
                 drop = true;
                 self.last_key_is_delete = true;
-            } else if epoch < self.task_config.watermark && self.watermark_can_see_last_key {
```

Comment:
It seems that the field `watermark_can_see_last_key` is not used anymore. We can remove the field.